### PR TITLE
Integrate RewriteAmpUrls transformer from toolbox library

### LINKED ISSUE #5581: Transformer: Implement self-hosting of AMP scripts
## Feature description

Depends on https://github.com/ampproject/amphtml/issues/25873.

Self-hosting will need to rely on fetching the required scripts from the AMP CDN to then serve locally via new rewrite rules that render out the cached responses. The AMP plugin would serve essentially as a proxy.

Some key benefits to doing this include:

* Fewer HTTP connections to obtain the JS critical for the page to load.
* Elimination of dependency on CDN which can be a seen as a privacy concern (nevertheless, see ampproject/amp-wp#4403).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by schlessera**
## Summary

This PR fetches the new changes from the toolbox library and integrates them with the plugin.

The ES Modules can be tested and profiled with the following plugin: https://gist.github.com/schlessera/4127ab30f4235f1acbc11e3eee77701a

Dependency for #4600 
Dependency for #5581
Related #5378

Depends on:

- [x] https://github.com/ampproject/amp-toolbox-php/pull/119
- [x] https://github.com/ampproject/amp-toolbox-php/pull/130

To opt-in to LTS versioning for AMP scripts, the following plugin code can be used:

```php
use AmpProject\Optimizer\Transformer\RewriteAmpUrls;
use AmpProject\Optimizer\Configuration\RewriteAmpUrlsConfiguration;

add_filter( 'amp_optimizer_config', static function ( $config ) {
	$config[ RewriteAmpUrls::class ][ RewriteAmpUrlsConfiguration::LTS ] = true;
	return $config;
} );
```

## Checklist

- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Review by westonruter — CHANGES_REQUESTED** [review]
I haven't yet looked into this in depth, but I'm seeing AMP validation errors:

image

For scripts, I'm seeing this:

```html
<script async nomodule src="https://cdn.ampproject.org/v0.js"></script>
<script src="https://cdn.ampproject.org/v0/amp-analytics-0.1.mjs" async="" custom-element="amp-analytics" type="module" crossorigin="anonymous"></script>
<script src="https://cdn.ampproject.org/v0/amp-bind-0.1.mjs" async="" custom-element="amp-bind" type="module" crossorigin="anonymous"></script>
<script src="https://cdn.ampproject.org/v0/amp-form-0.1.mjs" async="" custom-element="amp-form" type="module" crossorigin="anonymous"></script>
<script src="https://cdn.ampproject.org/v0/amp-install-serviceworker-0.1.mjs" async="" custom-element="amp-install-serviceworker" type="module" crossorigin="anonymous"></script>
<script src="https://cdn.ampproject.org/v0/amp-lightbox-0.1.mjs" async="" custom-element="amp-lightbox" type="module" crossorigin="anonymous"></script>
<script src="https://cdn.ampproject.org/v0/amp-mustache-0.2.mjs" async="" custom-template="amp-mustache" type="module" crossorigin="anonymous"></script>
<!-- dns-prefetch links, style[amp-custom], various other link elements... -->
<script async="" src="https://cdn.ampproject.org/v0.mjs" type="module" crossorigin="anonymous"></script>
```

It seems:

1. The `nomodule` versions of the AMP component scripts are missing, except for the AMP runtime.
2. The `v0.mjs` script is not located in the same spot as the `v0.js` script is located. I'd expect them to appear next to each other.

**Comment by schlessera** [comment]
Yes, that's why the PR is currently still in draft. Some component seems to strip the `nomodule` scripts at some point.

**Comment by schlessera** [comment]
The problem of the missing nomodules was due to the ordering of the default transformers, where `ReorderHead` came after `RewriteAmpUrls`. Switching these in the ordering solved the immediate issue.

However, this made me aware that the `ReorderHead` transformer uses the URL as array key, so it implicitly deduplicates based on the URL, which was the reason why the `nomodule` versions went missing. I've created an issue for this, as there might be other instances where multiple `<script>` elements share a same URL and need to pass through the `ReorderHead` transformer: https://github.com/ampproject/amp-toolbox-php/issues/125

**Comment by swissspidy** [comment]
Can't wait to leverage this in the Web Stories plugin as well 😻

File: includes/class-amp-theme-support.php [hunk 1]
```diff
@@ -2107,6 +2096,15 @@ private static function get_optimizer_configuration( $args ) {
...
 		$enable_ssr = apply_filters( 'amp_enable_ssr', $enable_ssr );
 
+		/**
+		 * Filter whether the AMP Optimizer should use ES modules for the runtime and extensions.
+		 *
+		 * @since 2.1.0
+		 *
+		 * @param bool $enable_esm Whether the AMP Optimizer should use ES modules for the runtime and extensions.
+		 */
+		$enable_esm = apply_filters( 'amp_enable_esm', $enable_esm );
```
Comment (westonruter) [inline includes/class-amp-theme-support.php]: Do you think a new `amp_enable_esm` filter for this is really needed? The `amp_optimize_config` can be used to turn it off:

```php
use AmpProject\Optimizer\Transformer\RewriteAmpUrls;
use AmpProject\Optimizer\Configuration\RewriteAmpUrlsConfiguration;

add_filter( 'amp_optimizer_config', static function ( $config ) {
	$config[ RewriteAmpUrls::class ][ RewriteAmpUrlsConfiguration::ESM_MODULES_ENABLED ] = false;
	return $config;
} );
```

Comment (schlessera) [reply includes/class-amp-theme-support.php]: No, it's not really needed, and I'm okay with removing it.

Comment (schlessera) [reply includes/class-amp-theme-support.php]: Updated my gist to work without the filter: https://gist.github.com/schlessera/4127ab30f4235f1acbc11e3eee77701a

File: includes/class-amp-theme-support.php [hunk 2]
```diff
@@ -2137,6 +2135,9 @@ private static function get_optimizer_configuration( $args ) {
 					Optimizer\Transformer\PreloadHeroImage::class => [
 						Optimizer\Configuration\PreloadHeroImageConfiguration::INLINE_STYLE_BACKUP_ATTRIBUTE => 'data-amp-original-style',
 					],
+					Optimizer\Transformer\RewriteAmpUrls::class => [
+						Optimizer\Configuration\RewriteAmpUrlsConfiguration::ESM_MODULES_ENABLED => $enable_esm,
+					],
```
Comment (westonruter) [inline includes/class-amp-theme-support.php]:
```suggestion
```

File: includes/class-amp-theme-support.php [hunk 3]
```diff
@@ -2107,6 +2096,15 @@ private static function get_optimizer_configuration( $args ) {
...
 
+		/**
+		 * Filter whether the AMP Optimizer should use ES modules for the runtime and extensions.
+		 *
+		 * @since 2.1.0
+		 *
+		 * @param bool $enable_esm Whether the AMP Optimizer should use ES modules for the runtime and extensions.
+		 */
+		$enable_esm = apply_filters( 'amp_enable_esm', $enable_esm );
+
```
Comment (westonruter) [inline includes/class-amp-theme-support.php]:
```suggestion
```

Comment (schlessera) [reply includes/class-amp-theme-support.php]: Can you edit this one to include one of the blank lines? Otherwise we'll end up with 2 of them...

File: includes/class-amp-theme-support.php (hunk 2 above)
Comment (westonruter) [reply includes/class-amp-theme-support.php]: Done in 3d1175a0a

File: includes/class-amp-theme-support.php (hunk 3 above)
Comment (westonruter) [reply includes/class-amp-theme-support.php]: Done in 3d1175a0a

**Comment by westonruter** [comment]
Note that I've changed the description to make it so it doesn't actually fix two issues: 

* #4600: There are additional changes needed to support release channels (namely to allow opting-in to beta/experimental which requires omission of `crossorigin=anonymous`).
* #5581: Self-hosting of scripts will require actually downloading the scripts and then once they are downloaded configure the `RewriteAmpUrls` transformer to point to them. Additional changes to AMP will be needed to support the plugin's needs, per https://github.com/ampproject/amphtml/issues/25873#issuecomment-812086561.

File: includes/class-amp-theme-support.php [hunk 4]
```diff
@@ -2114,6 +2100,7 @@ private static function get_optimizer_configuration( $args ) {
 				[
 					Optimizer\Transformer\AmpRuntimeCss::class,
 					Optimizer\Transformer\PreloadHeroImage::class,
+					Optimizer\Transformer\RewriteAmpUrls::class,
```
Comment (schlessera) [inline includes/class-amp-theme-support.php:2119]: Why was this added here? Disabling SSR should not have any effect on ESM, LTS or RTV, should it?

Comment (westonruter) [reply includes/class-amp-theme-support.php:2119]: This transformer requires `transformed` to be valid AMP. I added it here when trying to load a page with SSR disabled and it became invalid (and broken too).

**Review by westonruter — APPROVED** [review]
